Update docs for async/await syntax (#1576)
cocotb/cocotb: documentation/source/coroutines.rst

**********

Testbenches built using cocotb use coroutines. While the coroutine is executing
the simulation is paused. The coroutine uses the :keyword:`await` keyword to
block on another coroutine's execution or pass control of execution back to the
simulator, allowing simulation time to advance.

Typically coroutines :keyword:`await` a :class:`~cocotb.triggers.Trigger` object which
indicates to the simulator some event which will cause the coroutine to be woken
when it occurs.  For example:

.. code-block:: python3

    async def wait_10ns():
        cocotb.log.info("About to wait for 10 ns")
        await Timer(10, units='ns')
        cocotb.log.info("Simulation time has advanced by 10 ns")

> Coroutines may also wait other coroutines:

Reviewer: Either
```suggestion
Coroutines may also wait for other coroutines:
```
```suggestion
Coroutines may also wait upon other coroutines:
```
```suggestion
Coroutines may also wait on other coroutines:
```
```suggestion
Coroutines may also await other coroutines:
```

Reviewer 2: I like the version using await, otherwise someone could find it ambiguous why we don't use the keyword when it fits perfectly.

.. code-block:: python3

    async def wait_100ns():
        for i in range(10):
            await wait_10ns()

Coroutines can :keyword:`return` a value, so that they can be used by other coroutines.

.. code-block:: python3

    async def get_signal(clk, signal):
        await RisingEdge(clk)
        return signal.value

    async def check_signal_changes(dut):
        first = await get_signal(dut.clk, dut.signal)
        second = await get_signal(dut.clk, dut.signal)
        if first == second:
            raise TestFailure("Signal did not change")

> Coroutines can used with :func:`~cocotb.fork` for concurrent operation.

Reviewer 2: can be used

        # part of the user-generated "uart_tb" class
        cocotb.fork(tb.reset_dut(dut.rstn, 20))

        await Timer(10, units='ns')
        print("Reset is still active: %d" % dut.rstn)
        await Timer(15, units='ns')
        await("Reset has gone inactive: %d" % dut.rstn)

> Forked coroutines can used in an :keyword:`await` statement to block until the coroutine finishes.

Reviewer 2: ```suggestion
Forked coroutines can be used in an :keyword:`await` statement to block until the coroutine finishes.
```

.. code-block:: python3

    @cocotb.test()

>     async def test_count_edge_cycles(dut, period=1, clocks=6):

Reviewer 2: We could switch to period_ns while we are at it.

        cocotb.fork(Clock(dut.clk, period, units='ns').start())
        await RisingEdge(dut.clk)

>         timer = Timer(period + 10)

Reviewer 2: Use the units argument

        task = cocotb.fork(count_edges_cycles(dut.clk, clocks))
        count = 0
        expect = clocks - 1

        while True:

>             result = await cocotb.triggers.First(timer, task)

Reviewer: Perhaps we should import the short name like we do for the other triggers, as
```suggestion
            result = await First(timer, task)
```

        if not isclose(edge_time_ns, start_time_ns + 4.0):
            raise TestFailure("Expected a period of 4 ns")

.. _async_functions:

Async functions
===============

Python 3.5 introduces :keyword:`async` functions, which provide an alternative
syntax. For example:

.. code-block:: python3

    @cocotb.coroutine
    async def wait_10ns():
        cocotb.log.info("About to wait for 10 ns")
        await Timer(10, units='ns')
        cocotb.log.info("Simulation time has advanced by 10 ns")

To wait on a trigger or a nested coroutine, these use :keyword:`await` instead
of :keyword:`yield`. Provided they are decorated with ``@cocotb.coroutine``,
``async def`` functions using :keyword:`await` and regular functions using
:keyword:`yield` can be used interchangeable - the appropriate keyword to use
is determined by which type of function it appears in, not by the
sub-coroutine being called.

.. versionadded:: 1.4
    The :any:`cocotb.coroutine` decorator is no longer necessary for ``async def`` coroutines.
    ``async def`` coroutines can be used, without the ``@cocotb.coroutine`` decorator, wherever decorated coroutines are accepted,

>     including :keyword:`yield` statements and :any:`cocotb.fork`.

Reviewer: This section should be kept, but probably moved down to the section below

Author: I moved it to the bottom of the coroutine section since it applies to the `async` coroutines used in the examples.

.. note::
    It is not legal to ``await`` a list of triggers as can be done in
    ``yield``-based coroutine with ``yield [trig1, trig2]``. Use

>     ``await First(trig1, trig2)`` instead.

Reviewer: I think we should keep this note around somewhere too.

Author: It's in the documentation for `First`.

            else:
                break

Forked coroutines can be killed before they complete, forcing their completion before
they'd naturally end.

.. code-block:: python3

    @cocotb.test()
    async def test_different_clocks(dut):
        clk_1mhz   = Clock(dut.clk, 1.0, units='us')
        clk_250mhz = Clock(dut.clk, 4.0, units='ns')

        clk_gen = cocotb.fork(clk_1mhz.start())
        start_time_ns = get_sim_time(units='ns')
        await Timer(1, units='ns')
        await RisingEdge(dut.clk)
        edge_time_ns = get_sim_time(units='ns')
        if not isclose(edge_time_ns, start_time_ns + 1000.0):
            raise TestFailure("Expected a period of 1 us")

        clk_gen.kill()  # kill clock coroutine here

        clk_gen = cocotb.fork(clk_250mhz.start())
        start_time_ns = get_sim_time(units='ns')
        await Timer(1, units='ns')
        await RisingEdge(dut.clk)
        edge_time_ns = get_sim_time(units='ns')
        if not isclose(edge_time_ns, start_time_ns + 4.0):
            raise TestFailure("Expected a period of 4 ns")

> .. versionadded:: 1.4

Reviewer: Maybe out of scope, but thinking this should be
```suggestion
.. versionchanged:: 1.4
```

More details on this type of generator can be found in :pep:`525`.

> Generator-based coroutines

Reviewer: Likely will want a label on this, maybe

```suggestion
.. yield-syntax:

Generator-based coroutines
```

Or whatever the syntax for label markers is.

Author: Because it took me a while to figure it out, I'll mention it's:

`.. _ref_name:`

`:ref:'ref_name'`

Note the required underscore and lack of it when using it out.

Reviewer: Pretty sure the ref should be with backticks, not quotes.

Author: yeah, I just haven't figured out how to next backticks.

==========================

> Previous to Python 3.5 and the introduction of :keyword:`async` and :keyword:`await`, coroutines were implemented as wrappers around generators.

Reviewer: ```suggestion
Prior to Python 3.5, and the introduction of :keyword:`async` and :keyword:`await`, coroutines were implemented as wrappers around generators.
```

        yield Timer(10, units='ns')
        cocotb.log.info("Simulation time has advanced by 10 ns")

    @cocotb.coroutine
    def wait_100ns():
        for _ in range(10):
            yield wait_10ns()

You may see existing code that uses this syntax for coroutines, but do not worry, it is compatible with :keyword:`async` and :keyword:`await` coroutines.
Any place that will take :keyword:`async` coroutines, will also take :class:`~cocotb.decorators.coroutine`;
including :func:`~cocotb.fork`.
Anything that can be used in an :keyword:`await` statement can be used in a :keyword:`yield` statement inside of an generator-based coroutine;
including triggers like :class:`~cocotb.triggers.Timer`.

.. code-block:: python3

    async def wait_100ns():
        for _ in range(10):
            await wait_10ns()

.. code-block:: python3

    @cocotb.coroutine
    def clock(signal, half_period, half_period_units):
        signal <= 0
        while True:
            yield Timer(half_period, half_period_units)
            signal <= ~signal

    @cocotb.test()
    async def test(dut):
        cocotb.fork(clock(dut.clk, 5, 'ns'))
        await RisingEdge(dut.clk)

>     .. note:: This style is no longer recommended and support may someday be removed.

Reviewer: ```suggestion
.. note:: This style is no longer recommended and support may someday be removed.
```

More details on this type of generator can be found in :pep:`525`.

> .. _yield_syntax:

Reviewer: ```suggestion
.. _yield-syntax:
```
rst seems to prefer hyphenated targets

Reviewer: Can you put a link to here in the `First` docs?

Author: Is a note at the bottom sufficient? I'm not sure where else would be a natural place to add that.

Reviewer: Just add a link by the current "When using the old-style yield-based coroutines" note

.. _yield-syntax:

Generator-based coroutines
==========================

Prior to Python 3.5, and the introduction of :keyword:`async` and :keyword:`await`, coroutines were implemented as wrappers around generators.
Coroutine functions would be decorated with :class:`~cocotb.decorators.coroutine` and would use :keyword:`yield` to block on other coroutines or triggers.

.. code-block:: python3

    @cocotb.coroutine
    def wait_10ns():
        cocotb.log.info("About to wait for 10 ns")
        yield Timer(10, units='ns')
        cocotb.log.info("Simulation time has advanced by 10 ns")

    @cocotb.coroutine
    def wait_100ns():
        for _ in range(10):
            yield wait_10ns()

You may see existing code that uses this syntax for coroutines, but do not worry, it is compatible with async coroutines.
Any place that will accept async coroutines, will also accept generator-based coroutines;
including :func:`~cocotb.fork`.
Likewise, any object that can be used in an :keyword:`await` statement can also be used in a :keyword:`yield` statement while in a generator-based coroutine;
including triggers like :class:`~cocotb.triggers.Timer`.

> You may also see syntax like ``yield [triggerA, tiggerB, ...]``, this is syntax sugar for :class:`~cocotb.triggers.First`.

Reviewer: ```suggestion
You may also see syntax like ``yield [trigger_a, trigger_b, ...]``, which is syntactic sugar for :class:`~cocotb.triggers.First`.
```

Reviewer: Would still like to see a backlink from the First docs to this section.

including triggers like :class:`~cocotb.triggers.Timer`.

> You may also see syntax like ``yield [trigger_a, tigger_b, ...]``, which is syntactic sugar for :class:`~cocotb.triggers.First`.

Reviewer: I think this sentence should be below the example, since it seems to me the example is associated with the previous sentence.

Author: I also added a small example that is the example from `First`'s docstring, but using `yield []`.

    async def wait_100ns():
        for _ in range(10):
            await wait_10ns()

.. code-block:: python3

    @cocotb.coroutine
    def clock(signal, half_period, half_period_units):
        signal <= 0
        while True:
            yield Timer(half_period, half_period_units)
            signal <= ~signal

    @cocotb.test()
    async def test(dut):

>         cocotb.fork(clock(dut.clk, 5, 'ns'))

Reviewer 2: ```suggestion
        cocotb.fork(clock(dut.clk, 5, units='ns'))
```
for consistency

        await RisingEdge(dut.clk)

> .. note:: This style is no longer recommended and support may someday be removed.

Reviewer: I think this note would be best at the start of this section.

More details on this type of generator can be found in :pep:`525`.

.. _yield-syntax:

Generator-based coroutines
==========================

.. note:: This style is no longer recommended and support may someday be removed.

Prior to Python 3.5, and the introduction of :keyword:`async` and :keyword:`await`, coroutines were implemented as wrappers around generators.
Coroutine functions would be decorated with :class:`~cocotb.decorators.coroutine` and would use :keyword:`yield` to block on other coroutines or triggers.

.. code-block:: python3

    @cocotb.coroutine
    def wait_10ns():
        cocotb.log.info("About to wait for 10 ns")
        yield Timer(10, units='ns')
        cocotb.log.info("Simulation time has advanced by 10 ns")

    @cocotb.coroutine
    def wait_100ns():
        for _ in range(10):
            yield wait_10ns()

You may see existing code that uses this syntax for coroutines, but do not worry, it is compatible with async coroutines.

> Any place that will accept async coroutines, will also accept generator-based coroutines;

Reviewer 2: ```suggestion
Any place that will accept async coroutines will also accept generator-based coroutines;
```
?

Author: That would be correct. Imagine replacing the phrase "Any place that will accept async coroutines" with "They"; the comma would be out-of-place.

including :func:`~cocotb.fork`.
Likewise, any object that can be used in an :keyword:`await` statement can also be used in a :keyword:`yield` statement while in a generator-based coroutine;
including triggers like :class:`~cocotb.triggers.Timer`.

.. code-block:: python3

    async def wait_100ns():
        for _ in range(10):
            await wait_10ns()

.. code-block:: python3

    @cocotb.coroutine
    def clock(signal, half_period, half_period_units):
        signal <= 0
        while True:
            yield Timer(half_period, half_period_units)
            signal <= ~signal

>     @cocotb.test()

Reviewer: Perhaps worth mentioning that `test` objects can be generators too without the coroutine decorator.

    async def test(dut):
        cocotb.fork(clock(dut.clk, 5, units='ns'))
        await RisingEdge(dut.clk)

> You may also see syntax like ``yield [trigger_a, tigger_b, ...]``, which is syntactic sugar for :class:`~cocotb.triggers.First`.

Reviewer: ```suggestion
You may also see syntax like ``yield [trigger_a, trigger_b, ...]``, which is syntactic sugar for :class:`~cocotb.triggers.First`.
```

.. code-block:: python3

    t1 = Timer(10, units='ps')
    t2 = Timer(10, units='ps')

>     t_ret = yield[t1, t2]

Reviewer: ```suggestion
    t_ret = yield [t1, t2]
```

.. code-block:: python3

    @cocotb.coroutine
    def wait_10ns():
        cocotb.log.info("About to wait for 10 ns")
        yield Timer(10, units='ns')
        cocotb.log.info("Simulation time has advanced by 10 ns")

    @cocotb.coroutine
    def wait_100ns():
        for _ in range(10):
            yield wait_10ns()

You may see existing code that uses this syntax for coroutines, but do not worry, it is compatible with async coroutines.
Any place that will accept async coroutines will also accept generator-based coroutines;
including :func:`~cocotb.fork`.
Likewise, any object that can be used in an :keyword:`await` statement can also be used in a :keyword:`yield` statement while in a generator-based coroutine;
including triggers like :class:`~cocotb.triggers.Timer`.
Tests can also be generator-based coroutines.
Tests are not required to be decorated with :class:`~cocotb.decorators.coroutine` as the :class:`~cocotb.decorators.test` decorator will handle this case automatically.

.. code-block:: python3

    async def wait_100ns():
        for _ in range(10):

>             await wait_10ns()

Reviewer: ```suggestion
            # generator-based coroutines can still be awaited
            await wait_10ns()
```
Also, I think this example belongs after line 272.

Author: I'm not sure what line you meant, but I think you mean I could break up the previous paragraph and move the examples after every statement?

Reviewer: That's basically what I'm suggesting, yes - make it as clear as possible what each example is demonstrating.

Reviewer: Or at least, do it for these two blocks which are already separate.

Author: Done.

More details on this type of generator can be found in :pep:`525`.

.. _yield-syntax:

Generator-based coroutines
==========================

.. note:: This style is no longer recommended and support may someday be removed.

Prior to Python 3.5, and the introduction of :keyword:`async` and :keyword:`await`, coroutines were implemented as wrappers around generators.
Coroutine functions would be decorated with :class:`~cocotb.decorators.coroutine` and would use :keyword:`yield` to block on other coroutines or triggers.
You may see existing code that uses this syntax for coroutines, but do not worry, it is compatible with async coroutines.

Any object that can be used in an :keyword:`await` statement can also be used in a :keyword:`yield` statement while in a generator-based coroutine;
including triggers like :class:`~cocotb.triggers.Timer`.

.. code-block:: python3

    @cocotb.coroutine

>     def clock(signal, half_period, half_period_units):

Reviewer 2: Maybe call this eg myclock to make it clear even to a quick reader that this isn't our Clock class when used later

including :func:`~cocotb.fork`.

.. code-block:: python3

    @cocotb.coroutine
    def start_clock(clk):
        # generator-based coroutines can still be forked
        cocotb.fork(clock(clk, 5, units='ns'))
        yield RisingEdge(clk)

Async coroutines can be yielded in generator-based coroutines.

.. code-block:: python3

    async def detect_transaction(clk, valid):
        await RisingEdge(clk)
        while not valid.value:
            await RisingEdge(clk)

    @cocotb.coroutine
    def monitor(clk, valid, data):
        # async coroutines can be yielded
        yield detect_transaction(clk, valid)
        return data.value

Generator-based coroutines can also be awaited in async coroutines.

.. code-block:: python3

    async def check_incrementing(clk, valid, data):
        # generator-based coroutines can be awaited
        prev_count = await monitor()
        while True:
            count = await monitor()
            assert count == (prev_count + 1)
            prev_count = count

You may also see syntax like ``yield [trigger_a, trigger_b, ...]``, which is syntactic sugar for :class:`~cocotb.triggers.First`.

.. code-block:: python3

    @cocotb.coroutine

>     def run_for(coro, time, units):

Reviewer 2: You are not actually using time in the function body

Coroutine functions would be decorated with :class:`~cocotb.decorators.coroutine` and would use :keyword:`yield` to block on other coroutines or triggers.
You may see existing code that uses this syntax for coroutines, but do not worry, it is compatible with async coroutines.

Any object that can be used in an :keyword:`await` statement can also be used in a :keyword:`yield` statement while in a generator-based coroutine;
including triggers like :class:`~cocotb.triggers.Timer`.

.. code-block:: python3

    @cocotb.coroutine
    def clock(signal, half_period, half_period_units):
        signal <= 0
        while True:
            # in generator-based coroutines triggers are yielded
            yield Timer(half_period, half_period_units)
            signal <= ~signal

Likewise, any place that will accept async coroutines will also accept generator-based coroutines;
including :func:`~cocotb.fork`.

.. code-block:: python3

    @cocotb.coroutine
    def start_clock(clk):
        # generator-based coroutines can still be forked
        cocotb.fork(clock(clk, 5, units='ns'))
        yield RisingEdge(clk)

> Async coroutines can yielded in generator-based coroutines.

Reviewer: ```suggestion
Async coroutines can be yielded in generator-based coroutines.
```
